source/ring_buffer.c

```diff
@@ -40,32 +52,47 @@ void aws_ring_buffer_clean_up(struct aws_ring_buffer *ring_buf) {
     AWS_ZERO_STRUCT(*ring_buf);
 }
 
+bool check_atomic_ptr(struct aws_ring_buffer *const ring_buf, uint8_t *atomic_ptr) {
+    return (atomic_ptr >= ring_buf->allocation && atomic_ptr < ring_buf->allocation_end);
+}
+
 int aws_ring_buffer_acquire(struct aws_ring_buffer *ring_buf, size_t requested_size, struct aws_byte_buf *dest) {
+    AWS_PRECONDITION(aws_ring_buffer_is_valid(ring_buf));
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dest));
     if (requested_size == 0) {
+        AWS_POSTCONDITION(aws_ring_buffer_is_valid(ring_buf));
+        AWS_POSTCONDITION(aws_byte_buf_is_valid(dest));
         return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
     }
 
-    uint8_t *tail_cpy = aws_atomic_load_ptr(&ring_buf->tail);
-    uint8_t *head_cpy = aws_atomic_load_ptr(&ring_buf->head);
+    uint8_t *tail_cpy;
+    uint8_t *head_cpy;
+    AWS_ATOMIC_LOAD_PTR(tail_cpy, ring_buf, &ring_buf->tail);
+    AWS_ATOMIC_LOAD_PTR(head_cpy, ring_buf, &ring_buf->head);
 
     /* this branch is, we don't have any vended buffers. */
     if (head_cpy == tail_cpy) {
-        size_t ring_space = ring_buf->allocation_end - ring_buf->allocation;
-        size_t current_space = ring_buf->allocation_end - head_cpy;
+        size_t ring_space = ring_buf->allocation_end - ring_buf->allocation - 1;
+        size_t current_space = ring_buf->allocation_end - head_cpy - 1;
```

Comment:
`ring_space` and `current_space ` previously could lead to the `dest` buffer having access to an invalid unallocated memory, since at `aws_ring_buffer_init` function we allocated `size` bytes at `ring_buf->allocation` and `ring_buf->allocation_end` is set to be `ring_buf->allocation + size + 1`.

Reply:
this has been updated in the ring buffer code. Gonna move that over and then we can keep iterating on this.